MAINT: Refactor optimization methods to use scipy.stats.qmc

tupui opened the pull request:
This removes legacy QMC code in favour of `scipy.stats.qmc` functions introduced in #10844.

LHS is present in `scipy.optimize._differentialevolution.py` and Sobol' is implemented in pure python and used in `scipy.optimize._shgo.py`.

Also, currently the doc for `shgo` is using an incorrect number of samples for Sobol' as not being a power of 2.

andyfaff commented:
@tupui, what benefits are there to changing the LHS initialisation in differential evolution from what was there to the QMC implementation? As is, the code should be distributing the initial guess according to a Latin Hypersquare. The QMC won't make diffev work any faster, because it's a miniscule part of the total work. Moreover, we will also lose an amount of reproducibility against previous versions of scipy because the initialisation will be different for the same seed.

tupui commented:
> @tupui, what benefits are there to changing the LHS initialisation in differential evolution from what was there to the QMC implementation? As is, the code should be distributing the initial guess according to a Latin Hypersquare. The QMC won't make diffev work any faster, because it's a miniscule part of the total work. Moreover, we will also lose an amount of reproducibility against previous versions of scipy because the initialisation will be different for the same seed.

It is roughly the same implementation behind, but doing so would allow in the future to easily switch to any QMC for initialisation. Or maybe I should already do this here? So add other options for the parameter `init`. Going that direction, I would also suggest to default to Sobol' which has better properties than LHS which can be really bad.

andyfaff commented:
@rgommers, @ev-br, do we have a scipy policy on strict back compatibility with functions that use random numbers? In this PR changing the implementation of the `latinhypersquare` initialisation will mean that a `differential_evolution` minimisation done with different scipy versions will have different outputs for the same seed. There's no bug with the existing implementation, which would've justified the changes in this PR.

andyfaff commented:
@tupui, if `sobol` would provide a well distributed set of initial population guesses, then it might be nice to have a `sobol` option for `init`. But I wouldn't change it to make it the default option to start with, for the same reasons I mentioned in my previous comment.

ev-br commented:
Over at scipy.stats, we do change random streams once in a while without any backcompat notices. 
I've no informed opinion how severe this change can be in this regard.

rgommers commented:
I think we should document what the policy is. I'd be in favor of adopting the same policy as the new `numpy.random` one (for details, see [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html)): we try hard to not change results for a seed, but there's no guarantee for exact reproducibility between different SciPy versions - we reserve the right to make changes to fix bugs, and for sigficant performance or quality of random variates improvements.

tupui commented:
With the last commit I propose something toward adding other methods for initialization. If all this is OK, I will add tests and doc. Otherwise I can revert it OC. Just let me know what you think is best 😃

andyfaff commented:
@tupui, after reading through the link that Ralf sent through I think the best thing to do here is to revert the change to the `differential_evolution``init='latinhypersquare'`  keyword. The existing behaviour does not have bugs, it probably doesn't give significant benefit to make the changes in this PR, and it degrades reproducibility by giving different output from the same seed. That is, the code path for `init='latinhypersquare'` should give the same output before and after this PR.
However, a 'sobol' and 'halton' initialisation option would be welcomed. These should both be documented in the docstrings for `differential_evolution` and `DifferentialEvolutionSolver`. Note, the total population size is normally calculated as `popsize * len(x)`; this would conflict with the `2**m` behaviour for 'sobol'. Some thought should go into choosing the correct behaviour here. Should it be the closest power of 2? Note, the population should have at least 5 members, `m=2` is too small`.

tupui commented:
@andyfaff ok I reverted the original LHS code. I added some doc about Halton and Sobol'. To deal with the `ns=2**m` I am converting the population to the closest power of 2 (ceil and taking into account the min of 5).

As for leaving LHS as the default for compatibility, makes sense. But what about deprecating or using a warning to advice the user to move for another method?

tupui commented:
@Stefan-Endres what do you thing about these changes? I tried to not touch at anything else than the number of sample for Sobol' which should be a power of 2.

tupui commented:
* I am still fighting around with the number of samples. I noticed something odd in `iterate_delaunay`: is it normal that `self.nc += self.n` was before the sampling? To be it should be after otherwise for the first iteration we are sampling using more points. @Stefan-Endres, maybe you can help here?

* With the last commits, I update what I could in terms of number of samples for Sobol'. So at a new iteration, it is not using `self.nc` but it checks the internal state of the Sobol' generator to ask for the next `2**m` points (I still need to change `self.n` so that we have a correct `self.n, self.nc` but sounds hackish). It is not making a lot of sense to use Sobol' and sample something else than `2**m`.  If we must do this, then I would suggest we use Halton instead.

Please advise me on what you think I should do here @andyfaff @rgommers

rgommers commented on scipy/optimize/_shgo.py:
```diff
@@ -8,14 +8,14 @@
 
 
-def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
+def shgo(func, bounds, args=(), constraints=None, n=128, iters=1, callback=None,
```
This change is a little tricky - it does change the behavior for existing end user code. I expect that on average this increases the runtime, and it should also slightly increase the number of problems that get solved. Is it possible that this takes user code that was getting a valid solution before, and now fails to get a solution? That would be a problem - if so, changing to `n = None` and only changing to `2**m` points for `sobol` would address that.

tupui replied:
OK yes I can set it to `n=None`

rgommers commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -12,10 +12,11 @@
                                          NonlinearConstraint, LinearConstraint)
 from scipy.sparse import issparse
-
+from scipy.stats import qmc
```
This makes `optimize` depend on `stats`, which wasn't the case before. `stats` did already depend on `optimize` though. Apparently it doesn't create an import cycle, but it is slightly unfortunate. In other such situations, we've had to move the code that both submodules depend on to `scipy/_lib/`.

Something to look at a little more closely before merging.

rgommers commented:
With these kinds of change, there's always a question about whether there is a noticeable difference in performance (in terms of accuracy, harder than speed)? We have accuracy benchmarks in `benchmarks/benchmarks/optimize.py` for global optimizers, but unfortunately `shgo` isn't included there.

It's probably fine, since the change is not to the default method and relatively straightforward. But not quite sure - @andyfaff what do you think?

tupui commented:
> With these kinds of change, there's always a question about whether there is a noticeable difference in performance (in terms of accuracy, harder than speed)? We have accuracy benchmarks in `benchmarks/benchmarks/optimize.py` for global optimizers, but unfortunately `shgo` isn't included there.
> 
> It's probably fine, since the change is not to the default method and relatively straightforward. But not quite sure - @andyfaff what do you think?

I can make some benchmark on my side to check the speed.

What I am really concerned about is the way `sobol` is used. Currently it is used as a normal random generator. If we want minimal changes, using Halton instead would actually be a good option.

For Sobol', currently I am adding an exponential number of sample. This is not wanted I guess, so I will go ahed and use the scrambled version. This way we would be able to use a constant increase of number of point. Just have the 2**m constrain. Let me know @Stefan-Endres if this goes against theory.

tupui commented:
I am getting terrible results with the test function from the docstrings using the scrambled version. This is very interesting...

What I can do instead is to use the generator and fast forward it before any sampling. This way I can "emulate" randomization so that we can ask for a constant number of sample and keep the nice properties.

But this is puzzling. If someone knows about what is happening here that could help.

tupui commented:
With the last commits, I fixed the theoretical and practical issues with Sobol'.
So now we have a constant increase of points while having similar performance and respecting the theory behind the method. 

To have independent samples from the unscrambled version, I am skipping `self.iters_done * n` (with `n=2**m`) between every sampling.

Also, as a reminder, I changed the position of the update of the number of points as when specifying 64 points and 1 iteration, it was actually doing 64*2 points.

Stefan-Endres commented:
@tupui thank you for this initiative I think it would be great to refactor the random sampling !

>Actually I have to look more into this because when iterating we do not respect anymore the 2**m condition.

Unfortunately the typical use of `SHGO` is to repeat iterations (with fixed or variable sampling points generated per iteration) until a satisfactory solution is found. Therefore the 2**m recommendation will rarely be satisfied (and is usually not practical in higher dimensions). The other problem is that `SHGO` should also be able to handle non-linear constraints, so even though the sampling points are generated in a (stretched) hypercube, often the majority of these points will be discarded so in this case many of the low-discrepancy guarantees are void anyway.

>I am still fighting around with the number of samples. I noticed something odd in iterate_delaunay: is it normal that self.nc += self.n was before the sampling? To be it should be after otherwise for the first iteration we are sampling using more points. @Stefan-Endres, maybe you can help here?

This is a known bug that resulted in an extra iteration. We have fixed this in the upstream repository and I will open a PR to `scipy` very soon (all that's left is to update the documentation). In the latest version we have streamlined the code a lot so that the sampling points is counted in the same way that the `simplicial` method counts (and therefore you do not need to make any changes, because any sampling method will work correctly in the new version).

>It is not making a lot of sense to use Sobol' and sample something else than 2**m. If we must do this, then I would suggest we use Halton instead.

I think you are absolutely correct from a statistical perspective, from an optimisation perspective this is not practical in higher dimensions because of intermittent convergence criteria. I think that adding `Halton` above `Sobol` is a good idea. Any low discrepancy sampling method should work. My idea was to support any sequence that can be generated since some use cases have particular needs.

>Another possibility would be to use the scrambled version of Sobol' (which in any case we should: I am just keeping the base sequence here in case there is any theoretical reason I am not aware of) so that we don't add points exponentially.

This would be preferable. My question is how different would this be from the current Sobol sequence generated by SHGO? I have often get very different performance by using different seeds for the sequence.

@rgommers 
>With these kinds of change, there's always a question about whether there is a noticeable difference in performance (in terms of accuracy, harder than speed)? We have accuracy benchmarks in benchmarks/benchmarks/optimize.py for global optimizers, but unfortunately shgo isn't included there.

I have this in my offline library and just need to clean the code, I will try to open a separate PR as soon as possible.

tupui commented:
Thank you for the clarifications @Stefan-Endres 

I also noticed these differences in terms of results with the seed.

The main difference with the new implementation of Sobol' is that it is radically faster (cython), can be scrambled and also we checked the convergence.

I am not sure I understand when you say that you can have a varying number of points. When setting a `n` I could see it was using it all the time. The way I understand the algorithm is that it uses some batches of points. And at least for the generation of batches, we have a coherent sampling. Having some good coverage here would seems to be wanted, meaning we could benefit from QMC

So from what you say, I could adapt the code to accept also Halton. And I will use then the scrambled version in both cases. I will just need to adapt the number of points for the test cases due to this seeding variation.

As you wanted to work on this, shall I add you to my fork so that you could commit changes there? It might be easier if we both work on the same branch. What do you think?

Stefan-Endres commented:
>The main difference with the new implementation of Sobol' is that it is radically faster (cython), can be scrambled and also we checked the convergence. Still, I also noticed these differences in terms of results with the seed. I guess it's really a matter of the function and the "luck" we had with the bare unscrambled sequence. That's why I was asking if you had maybe some specific properties you needed from the base sequence. Like having the point (0, ..., 0).

If possible sampling the centre of the hypercube (0.5, 0.5, ..., 0.5) usually produces better performance (especially when the optimisation bounds are infinite or unknown), but I can connect this point manually in `shgo` later as long as the unittests are passing. I think in general there won't be any major change in benchmark performance. As long as the unittests are passing it would be ideal to merge this code with scipy before the next PR for shgo. Since we fixed the issue with an extra iteration run, the overall performance of the algorithm should be increased anyway if we coordinate to release this in the same scipy version.

>I am not sure I understand when you say that you can have a varying number of points. 

There is also an option (```options={`infty_constraints':False}```) with Sobol sampling to keep generating sampling points until a feasible number of points is found. For example, if the algorithm generated `n=64` points in a new iteration and 10 points are infeasible then `shgo` will generate another 10 points from QMC and this loop repeats until there are 64 feasible points. This is vital in cases where the user needs to keep memory usage low by not adding these points to the triangulation.

>When setting a n I could see it was using it all the time. The way I understand the algorithm is that it uses some batches of points. And at least for the generation of batches, we have a coherent sampling. Having some good coverage here would seems to be wanted, meaning we could benefit from QMC.

If you do not specify any constraints and the objective function is well behaved then it will work like you expect with `n = 2**m` generated with every call to QMC. The default behaviour of `shgo` is to not discard infeasible points and then the calls to QMC should again always be `n = 2**m`.

>So from what you say, I could adapt the code to accept also Halton. And I will use then the scrambled version in both cases. I will just need to adapt the number of points for the test cases due to this seeding variation.

Yes, I suspect that Halton would deliver better performance in general anyway, and ideally we can keep `sobol` for back compatibility together with the recommendation to use 2**m sampling points in the documentation. Since `sobol` is not the default sampling method anyway we would not need to make any API changes.

>As you wanted to work on this, shall I add you to my fork so that you could commit changes there? It might be easier if we both work on the same branch. What do you think?

Yeah, I think the best approach should be to implement Halton in this fork, then I will work on making `sobol` back compatible to ensure all the tests pass. It will be a bit of a merge conflict with the newer `shgo` code, I think it will unfortunately take a while to review that PR, so it will be best to keep the new routine separate and deal with that merge later after this one has been merged.

tupui commented:
Ok perfect then I will add Halton and we keep Sobol' for back compatibility. I will update the doc as well. I will do this early next week so that hopefully we can move forward with this.

tupui commented:
@Stefan-Endres I have a first version if you would like to review. Thank you!

Stefan-Endres commented:
@tupui thank you for this, this looks perfect and everything is running. I will pull this into the upstream library as well. I look forward to posting the benchmarks using ``halton`` sampling.

tupui commented:
I got this error in the CI

```python
scipy/optimize/optimize.py:282: in _check_clip_x
    warnings.warn("Values in x were outside bounds during a "
E   RuntimeWarning: Values in x were outside bounds during a minimize step, clipping to bounds
        bounds     = (array([-6., -5.]), array([6., 5.]))
        x          = array([1.69731632, 5.        ])
```

Not 100% sure how this could be related to my changes, but it keeps happening. Anyway, for now I just suppressed the warning during the test. Looking at `_check_clip_x`, I would actually change the sign checks from `(x < bounds[0]).any() or (x > bounds[1]).any()` to `(x <= bounds[0]).any() or (x >= bounds[1]).any()`. I can do another PR if you think this is correct. @rgommers?

rgommers commented:
> Not 100% sure how this could be related to my changes, but it keeps happening. Anyway, for now I just suppressed the warning during the test. Looking at `_check_clip_x`, I would actually change the sign checks from `(x < bounds[0]).any() or (x > bounds[1]).any()` to `(x <= bounds[0]).any() or (x >= bounds[1]).any()`. I can do another PR if you think this is correct. @rgommers?

Ugh, bounds handling is quite complicated in these functions and there has been lots of churn plus some open PRs. I have no idea what is correct without diving in. Ignoring the warning for now seems sensible.

andyfaff commented:
> I would actually change the sign checks from `(x < bounds[0]).any() or (x > bounds[1]).any()` to `(x <= bounds[0]).any() or (x >= bounds[1]).any()`

If a value is on the bounds it hasn't exceeded them and should be allowed. Therefore, the inequalities should be less/greater than.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -123,17 +124,25 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
-        covered. Use of an array to specify a population subset could be used,
+        maximize coverage of the available parameter space.
+
+        'sobol' and 'halton' are superior alternatives and maximize even more
```
It's not clear to me why they are superior alternatives. Can you briefly explain why, or give a reference? Use of the term superior should be justified.

tupui replied:
I will refer to `scipy.stats.qmc` for more details.

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -578,13 +595,20 @@ def __init__(self, func, bounds, args=(),
         # least 5 long
         self.num_population_members = max(5, popsize * self.parameter_count)
 
+        if isinstance(init, str) and (init == 'sobol'):
```
why not just have `if init == 'sobol'`?

tupui replied:
Indeed!

tupui replied:
Actually I remember why I did this. Because if `init` is not a str but an array, the comparison fails. I will add a comment in the code.
```python
>       if init == 'sobol':
E       FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
```

andyfaff commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -578,13 +595,20 @@ def __init__(self, func, bounds, args=(),
 
+        if isinstance(init, str) and (init == 'sobol'):
+            # must be Ns=2**m for Sobol'
+            n_s = int(2 ** np.ceil(np.log2(self.num_population_members)))
```
This has the effect of overriding the docstring:

> The population has popsize * len(x) individuals`.

 Perhaps a note should be added to that effect somewhere.

tupui replied:
I added a note in the description of `popsize` and changed the wording on the `init` part.

tupui commented:
> If a value is on the bounds it hasn't exceeded them and should be allowed. Therefore, the inequalities should be less/greater than (pretty sure that code shouldn't be changed).

Yes it should be allowed, hence I don't get why we should clip. In any case, this would just remove the warning in case we are on the bounds.

andyfaff commented on scipy/optimize/tests/test__dual_annealing.py:
```diff
@@ -178,10 +178,14 @@ def test_local_search_option_bounds(self):
-            local_search_options={"method": "SLSQP", "bounds": bounds})
+
+        with np.testing.suppress_warnings() as sup:
+            sup.record(RuntimeWarning, "Values in x were outside bounds ")
```
The reason this occurs is because SLSQP sometimes suggests trial solutions that are not with the bounds, i.e.:
`lb <= x <= ub` is not obeyed. It's only outside by a small amount, 1 or 2 ULP. This is an issue with the SLSQP Fortran code. The solution we came up with is to clip to the bounds; values on the bounds are acceptable.

andyfaff commented on scipy/optimize/tests/test__differential_evolution.py:
```diff
@@ -1029,7 +1029,9 @@ def c1(x):
         assert_(np.all(res.x <= np.array(bounds)[:, 1]))
 
-    def test_L5(self):
+    @pytest.mark.parametrize("random_generator",
```
probably best not to try this on one of the `test_Lx` series. They can take a long time to complete as it is.
The addition of the extra keywords should tested along the lines of `test_population_initiation` (L499-508).

andyfaff commented:
The warning is only emitted if a minimizer (SLSQP and TNC are the offenders) suggests a trial solution that is outside the bounds. In that situation the trial solution needs to be clipped. A warning is generated because this interferes with the operation of the minimizer.

If anything the warning is superfluous, but the inequality testing to see whether clipping needs to occur is correct. The testing is done because `np.clip` is reasonably expensive and it's actually quicker to do the check first.

tupui commented:
> The warning is only emitted if a minimizer (SLSQP and TNC are the offenders) suggests a trial solution that is outside the bounds. In that situation the trial solution needs to be clipped. A warning is generated because this interferes with the operation of the minimizer.
> 
> If anything the warning is superfluous, but the inequality testing to see whether clipping needs to occur is correct. The testing is done because `np.clip` is reasonably expensive and it's actually quicker to do the check first.

I think we have a misunderstanding here.

The way I understand the code is that: if one point is inside of the domain and touching the boundary, the clipping will happen. I am suggesting to change the test and allow equality so that touching the boundary does not require any clipping. In this case, the output would be identical, the clipping would not change anything as we touch and don't exceed the boundary. So the warning and the clipping are useless in this case.

andyfaff commented:
> The way I understand the code is that: if one point is inside of the domain and touching the boundary, the clipping will happen.

Your understanding is incorrect.



`if (x < bounds[0]).any() or (x > bounds[1]).any()` could be rewritten in English as: `if any of the points in *x* are less than the lower bound, or if any of the points in *x* are greater then the upper bound, then clip`.

At the moment touching the boundary does not do any clipping.

```
import numpy as np
import warnings

def _check_clip_x(x, bounds):
    if (x < bounds[0]).any() or (x > bounds[1]).any():
        warnings.warn("Values in x were outside bounds during a "
                      "minimize step, clipping to bounds", RuntimeWarning)
        x = np.clip(x, bounds[0], bounds[1])
        return x

    return x

>>> x = np.array([1.0])
>>> bounds = (np.array([0.0]), np.array([1.0]))
>>> _check_clip_x(x, bounds)
array([1.])
>>> x[0] = 0.0
array([0.])
>>> x[0] = 1.00000000001
>>> _check_clip_x(x, bounds)
<ipython-input-5-100bc10c0bcc>:6: RuntimeWarning: Values in x were outside bounds during a minimize step, clipping to bounds
  warnings.warn("Values in x were outside bounds during a "
array([1.])
```

tupui commented:
> > The way I understand the code is that: if one point is inside of the domain and touching the boundary, the clipping will happen.
> 
> Your understanding is incorrect.
> 
> `if (x < bounds[0]).any() or (x > bounds[1]).any()` could be rewritten in English as: `if any of the points in *x* are less than the lower bound, or if any of the points in *x* are greater then the upper bound, then clip`.
> 
> At the moment touching the boundary does not do any clipping.
> 
> ```
> import numpy as np
> import warnings
> 
> def _check_clip_x(x, bounds):
>     if (x < bounds[0]).any() or (x > bounds[1]).any():
>         warnings.warn("Values in x were outside bounds during a "
>                       "minimize step, clipping to bounds", RuntimeWarning)
>         x = np.clip(x, bounds[0], bounds[1])
>         return x
> 
>     return x
> 
> >>> x = np.array([1.0])
> >>> bounds = (np.array([0.0]), np.array([1.0]))
> >>> _check_clip_x(x, bounds)
> array([1.])
> >>> x[0] = 0.0
> array([0.])
> >>> x[0] = 1.00000000001
> >>> _check_clip_x(x, bounds)
> <ipython-input-5-100bc10c0bcc>:6: RuntimeWarning: Values in x were outside bounds during a minimize step, clipping to bounds
>   warnings.warn("Values in x were outside bounds during a "
> array([1.])
> ```

Oh right, now I see my mistake! Sorry I got confused 🙇‍♂️

tupui commented:
@andyfaff Thanks for the review, I updated the PR with your suggestions.

mdhaber commented:
@tupui my understanding per our discussion [here](https://github.com/scipy/scipy/pull/13441/files#r584380772) is that when this merges, `shgo` will round `n` up to the next power of two when using Sobol. If that is correct, and if gh-13441 merges before this, please adjust the documentation of `barnard_exact` accordingly.

tupui commented:
@rgommers As you expected, I encountered more circular imports now... I am not sure how to deal with these. For now, I imported `scipy.stats.qmc` where it was used. I had an extra issue due to the recent merge of `barnard_exact` as it introduced a dependency to `shgo`.

rgommers commented:
> @rgommers As you expected, I encountered more circular imports now... I am not sure how to deal with these. For now, I imported `scipy.stats.qmc` where it was used. I had an extra issue due to the recent merge of `barnard_exact` as it introduced a dependency to `shgo`.

That seems okay for now. It's not easy to refactor, given that you're needing `Sobol` and `Halton` it's not a matter of putting some common bits in `scipy._lib`, we'd have to move a significant fraction of `_qmc.py` which doesn't look appealing.

tupui commented:
> That seems okay for now. It's not easy to refactor, given that you're needing `Sobol` and `Halton` it's not a matter of putting some common bits in `scipy._lib`, we'd have to move a significant fraction of `_qmc.py` which doesn't look appealing.

Yes, let see if other modules would need it. For now I guess it's going to be it, but maybe at some point someone wants to use QMC to integrate or interpolate. In that case we might need to reconsider.

Is there anything else to do here or can we go ahead and merge? @mdhaber and @andyfaff, is it ok on your side now?

mdhaber commented:
This addresses the concern about documentation of the `n` parameter noted in gh-13441; thanks.

tupui commented:
@rgommers @mdhaber If you are OK and there are no more comments, I suggest I merge this before the WE.

mdhaber commented:
Sorry, I can't comment on the rest of this PR. I can only confirm that it addresses the concern in gh-13441.

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -77,9 +77,14 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
     iters : int, optional
-        Number of iterations used in the construction of the simplicial complex.
+        Number of iterations used in the construction of the simplicial
+        complex. Default is 1.
```
The default is already specified in the function signature, not really needed here.

tupui replied:
But I thought we must add the default in the description as well?? We do this almost everywhere and I remember reviewers asking for this in the past on another PR.

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -88,7 +93,7 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
 
             * method : str
-                The minimization method (e.g. ``SLSQP``).
+                The minimization method (Default is ``SLSQP``).
```
The minimization method, the default is ``SLSQP``.

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -305,58 +303,58 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
-    `shgo` has two built-in low discrepancy sampling sequences. First, we will
-    input 30 initial sampling points of the Sobol sequence:
+    `shgo` has built-in low discrepancy sampling sequences. First, we will
+    input 64 initial sampling points of the Sobol' sequence:
```
no need for the apostrophe there.

tupui replied:
This is the name of the method which has an apostrophe `Sobol'`. I added some italic to help the user.

andyfaff replied:
I think it should either be:

> of Sobol's sequence

or

> of the Sobol sequence

but definitely not `Sobol'`

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -305,58 +303,58 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
-    default of 1 to 5. This gives us 60 x 5 = 300 initial sampling points.
+    number of sampling points to 64 and the number of iterations from the
+    default of 1 to 5. Using ``simplicial`` this would have gives us
+    64 x 3 = 192 initial sampling points.
```
64 * 5 is 320.

andyfaff replied:
Or are you increasing the number of iterations from 1 to 3?

tupui replied:
Was 3 indeed.

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -305,58 +303,58 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
     In the first case the promising points contained in the minimiser pool
-    is processed only once. In the latter case it is processed every 60 sampling
-    points for a total of 3 times.
+    is processed only once. In the latter case it is processed every 64
```
is --> are

andyfaff commented on scipy/optimize/tests/test__shgo.py:
```diff
@@ -351,16 +351,16 @@ def test_f4_sobol(self):
         # run_test(test4_1, n=800)
         options = {'infty_constraints': False}
-        run_test(test4_1, n=990, options=options)
+        run_test(test4_1, n=2048, options=options)
```
Do these increase test run times?

andyfaff commented on scipy/optimize/tests/test__shgo.py:
```diff
@@ -624,20 +624,10 @@ def test_12_sobol_inf_cons(self):
         shgo(test1_2.f, test1_2.bounds, n=1, iters=None,
              options=options, sampling_method='sobol')
 
-    def test_13_high_sobol(self):
```
Why is this test removed?

tupui replied:
Because its goal was to check that the underlying code for `Sobol'` is called with dim > 40 as there was 2 implementations. One for < 40 and another for > 40. This is not relevant anymore as all of this is checked directly in our `stats.qmc` submodule.

andyfaff commented on scipy/stats/_hypotests.py:
```diff
@@ -696,10 +696,11 @@ def barnard_exact(table, alternative="two-sided", pooled=True, n=32):
-        enough to reach good precision. More points comes at performance cost.
+        method. Note that this argument will automatically be converted to
+        the next higher power of 2 since `scipy.stats.qmc.Sobol` is used to
+        select sample point. Default is 32. Must be positive. In most cases,
```
point --> points.
No need to mention default, it's in the function signature.

andyfaff reviewed (commented):
I've gone through all the files except `_shgo`. I'm not an expert on that file.

andyfaff commented:
@tupui, in general it's good practice that PR's are merged by someone different to the person that created it.

I don't believe it's a hard and fast rule, the exceptions are if the PR is fairly simple and straightforward. This PR is a bit more complicated.

tupui commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -123,17 +123,25 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
+
+        'sobol' and 'halton' are superior alternatives and maximize even more
+        the parameter space. 'sobol' will always use an initial population
+        which is a power of 2. 'halton' has no requirements but is a bit less
```
I rephrase, let me know if it's better.

tupui commented:
> I've gone through all the files except `_shgo`. I'm not an expert on that file.

Thanks @andyfaff for the review. I addressed all the points. For `shgo` itself, @Stefan-Endres approved earlier in the discussion. Apparently, he has some other improvements that he plan to do in a follow up PR.

andyfaff commented on scipy/optimize/_shgo.py:
```diff
@@ -305,58 +303,58 @@ def shgo(func, bounds, args=(), constraints=None, n=100, iters=1, callback=None,
-    number of sampling points to 60 and the number of iterations from the
-    default of 1 to 5. This gives us 60 x 5 = 300 initial sampling points.
+    number of sampling points to 64 and the number of iterations from the
+    default of 1 to 3. Using ``simplicial`` this would have gives us
```
'this would have given'

tupui commented:
@andyfaff if you have time, could you have another look at it?

tupui commented:
Friendly remainder @andyfaff. Thanks in advance.

tupui commented:
@rgommers if you have some time, could you have a look? It should be ready as it has been reviewed.

tupui commented:
@larsoner if you have some time, could you have a look?

tupui commented:
@mdhaber this one of mine is still pending. If you have some time to help ;)

mdhaber commented:
@tupui these days (until the summer probably) I would prefer to stick to stats. How about I take a look at gh-13471, instead?

tupui commented:
> @tupui these days (until the summer probably) I would prefer to stick to stats. How about I take a look at gh-13471, instead?

Sure thanks @mdhaber!

larsoner reviewed (approved):
> @larsoner if you have some time, could you have a look?

Not really in my domain (more signals/linalg) but given that @andyfaff looked, his comments seem to have been addressed, and from what I can tell it makes sense, I think we can merge this. Let's give folks who have been pinged a few more days -- @tupui can you ping me on Monday to merge this if I haven't already?

tupui commented:
> > @larsoner if you have some time, could you have a look?
> 
> Not really in my domain (more signals/linalg) but given that @andyfaff looked, his comments seem to have been addressed, and from what I can tell it makes sense, I think we can merge this. Let's give folks who have been pinged a few more days -- @tupui can you ping me on Monday to merge this if I haven't already?

Thank you 😃 . Sure I will ping you if this is not merged by then.

tupui commented:
Friendly reminder @larsoner  😄  Thanks.

larsoner commented:
Thanks @tupui !

tupui commented:
Thanks Eric!

@Stefan-Endres it's merged, you can now propose your enhancements! Feel free to ping me.
